File: backend/services/facades/player_facade.py
"""
Player state facade - FS-first player state management.

This facade owns the FS↔DB sync boundary for player state:
- FS (player.yaml) is the authoritative source
- DB (PlayerState) is an optional cache for fast reads/polling

All mutations go through this facade to ensure consistency.
"""

import logging
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Dict, List, Optional

from domain.entities.world_models import PlayerState as FSPlayerState
from domain.services.player_rules import (
    InventoryItem,
    apply_stat_changes,
    merge_inventory_item,
    remove_inventory_item,
)
from domain.services.player_state_serializer import PlayerStateSerializer

from services.player_service import PlayerService

if TYPE_CHECKING:
    from sqlalchemy.ext.asyncio import AsyncSession

logger = logging.getLogger("PlayerFacade")


@dataclass
class StatChangeResult:
    """Result of a stat change operation."""

    old_stats: Dict[str, int]
    new_stats: Dict[str, int]
    changes_applied: Dict[str, int]


@dataclass
class InventoryChangeResult:
    """Result of an inventory change operation."""

    success: bool
    inventory: List[Dict[str, Any]]
    message: str


@dataclass
class TimeAdvanceResult:
    """Result of a time advancement operation."""

    old_time: Dict[str, int]
    new_time: Dict[str, int]
    minutes_advanced: int


class PlayerFacade:
    """
    FS-first player state management facade.

    Usage (FS-only):
        facade = PlayerFacade(world_name)
        result = facade.update_stats({"health": -10, "experience": 50})

    Usage (with DB sync for polling):
        facade = PlayerFacade(world_name, db=db_session, world_id=123)
        result = await facade.update_stats({"health": -10})  # Syncs to DB
    """

    def __init__(
        self,
        world_name: str,
        db: Optional["AsyncSession"] = None,
        world_id: Optional[int] = None,
    ):
        """
        Initialize the facade for a specific world.

        Args:
            world_name: Name of the world (maps to worlds/{world_name}/)
            db: Optional DB session for cache sync
            world_id: Optional world ID for DB operations (required if db provided)
        """
        self.world_name = world_name
        self.db = db
        self.world_id = world_id

    def _load_state(self) -> Optional[FSPlayerState]:
        """Load current player state from FS."""
        return PlayerService.load_player_state(self.world_name)

    def _save_state(self, state: FSPlayerState) -> None:
        """Save player state to FS."""
        PlayerService.save_player_state(self.world_name, state)

    def _load_stat_definitions(self) -> Dict[str, Any]:
        """Load stat definitions from FS."""
        return PlayerService.load_stat_definitions(self.world_name)

    async def _sync_to_db(self, state: FSPlayerState) -> None:
        """
        Sync FS state to DB cache (for polling reads).

        This is a fire-and-forget operation - DB is just a cache.
        """
        if not self.db or not self.world_id:
            return

        try:
            from infrastructure.database import models
            from infrastructure.database.connection import serialized_write
            from sqlalchemy.future import select

            result = await self.db.execute(
                select(models.PlayerState).where(models.PlayerState.world_id == self.world_id)
            )
            db_state = result.scalar_one_or_none()

            if db_state:
                db_state.stats = PlayerStateSerializer.serialize_stats(state.stats)
                db_state.inventory = PlayerStateSerializer.serialize_inventory(state.inventory)
                async with serialized_write():
                    await self.db.commit()
                logger.debug(f"📥 Synced player state to DB cache for world {self.world_id}")
        except Exception as e:
            logger.warning(f"Failed to sync player state to DB: {e}")

    # =========================================================================
    # Stats Operations (FS-first)
    # =========================================================================

    async def update_stats(self, changes: Dict[str, int]) -> Optional[StatChangeResult]:
        """
        Update player stats (FS-first, then DB cache).

        Args:
            changes: Dict of stat_name -> delta (positive or negative)

        Returns:
            StatChangeResult with old/new values, or None if state not found
        """
        state = self._load_state()
        if not state:
            logger.warning(f"Player state not found for world: {self.world_name}")
            return None

        old_stats = state.stats.copy()
        stat_defs = self._load_stat_definitions()

        # Apply changes with clamping (using domain rules)
        new_stats = apply_stat_changes(state.stats, changes, stat_defs)
        state.stats = new_stats

        # Save to FS (authoritative)
        self._save_state(state)
        logger.info(f"📊 Updated stats for {self.world_name}: {changes}")

        # Sync to DB cache
        await self._sync_to_db(state)

        return StatChangeResult(
            old_stats=old_stats,
            new_stats=new_stats,
            changes_applied=changes,
        )

    def get_stats(self) -> Dict[str, int]:
        """Get current player stats from FS."""
        state = self._load_state()
        return state.stats if state else {}

    # =========================================================================
    # Inventory Operations (FS-first)
    # =========================================================================

    async def add_item(
        self,
        item_id: str,
        name: str,
        quantity: int = 1,
        description: Optional[str] = None,
        properties: Optional[Dict[str, Any]] = None,
    ) -> Optional[InventoryChangeResult]:
        """
        Add an item to inventory (FS-first, then DB cache).

        If item with same ID exists, quantity is added to existing.

        Args:
            item_id: Unique item identifier
            name: Display name
            quantity: Amount to add (default 1)
            description: Optional item description
            properties: Optional item properties dict

        Returns:
            InventoryChangeResult, or None if state not found
        """
        state = self._load_state()
        if not state:
            logger.warning(f"Player state not found for world: {self.world_name}")
            return None

        # Create domain item and merge (using domain rules)
        item = InventoryItem(
            id=item_id,
            name=name,
            quantity=quantity,
            description=description,
            properties=properties,
        )
        new_inventory = merge_inventory_item(state.inventory, item)
        state.inventory = new_inventory

        # Save to FS (authoritative)
        self._save_state(state)
        logger.info(f"📦 Added item to {self.world_name}: {name} x{quantity}")

        # Sync to DB cache
        await self._sync_to_db(state)

        return InventoryChangeResult(
            success=True,
            inventory=new_inventory,
            message=f"Added {name} x{quantity}",
        )

    async def remove_item(
        self,
        item_id: str,
        quantity: int = 1,
    ) -> Optional[InventoryChangeResult]:
        """
        Remove an item from inventory (FS-first, then DB cache).

        Args:
            item_id: Item ID to remove
            quantity: Amount to remove (default 1)

        Returns:
            InventoryChangeResult with success status, or None if state not found
        """
        state = self._load_state()
        if not state:
            logger.warning(f"Player state not found for world: {self.world_name}")
            return None

        # Remove item (using domain rules)
        new_inventory, success, remaining = remove_inventory_item(state.inventory, item_id, quantity)

        if not success:
            return InventoryChangeResult(
                success=False,
                inventory=state.inventory,
                message=f"Cannot remove {quantity} of item {item_id} (only {remaining} available)",
            )

        state.inventory = new_inventory

        # Save to FS (authoritative)
        self._save_state(state)
        logger.info(f"📦 Removed item from {self.world_name}: {item_id} x{quantity}")

        # Sync to DB cache
        await self._sync_to_db(state)

        return InventoryChangeResult(
            success=True,
            inventory=new_inventory,
            message=f"Removed {item_id} x{quantity}",
        )

    def get_inventory(self, resolved: bool = True) -> List[Dict[str, Any]]:
        """Get current inventory from FS.

        Args:
            resolved: If True, resolve item references to full data.
                     If False, return raw references from player.yaml.

        Returns:
            List of inventory items
        """
        if resolved:
            return PlayerService.get_resolved_inventory(self.world_name)
        else:
            state = self._load_state()
            return state.inventory if state else []

    # =========================================================================
    # Time Operations (FS-first)
    # =========================================================================

    async def advance_time(self, minutes: int) -> Optional[TimeAdvanceResult]:
        """
        Advance in-game time (FS-first, then DB cache).

        Handles day rollover at 24:00 (1440 minutes).

        Args:
            minutes: Number of minutes to advance

        Returns:
            TimeAdvanceResult with old/new time, or None if state not found
        """
        if minutes <= 0:
            return None

        state = self._load_state()
        if not state:
            logger.warning(f"Player state not found for world: {self.world_name}")
            return None

        old_time = state.game_time.copy()

        # Calculate new time
        total_minutes = state.game_time["hour"] * 60 + state.game_time["minute"] + minutes
        days_passed = total_minutes // 1440  # 1440 minutes per day
        remaining_minutes = total_minutes % 1440

        new_hour = remaining_minutes // 60
        new_minute = remaining_minutes % 60
        new_day = state.game_time["day"] + days_passed

        state.game_time = {
            "hour": new_hour,
            "minute": new_minute,
            "day": new_day,
        }

        # Save to FS (authoritative)
        self._save_state(state)
        logger.info(
            f"⏰ Advanced time for {self.world_name}: +{minutes}min -> {new_hour:02d}:{new_minute:02d} Day {new_day}"
        )

        # Sync to DB cache
        await self._sync_to_db(state)

        return TimeAdvanceResult(
            old_time=old_time,
            new_time=state.game_time,
            minutes_advanced=minutes,
        )

    def get_game_time(self) -> Dict[str, int]:
        """Get current game time from FS."""
        state = self._load_state()
        return state.game_time if state else {"hour": 8, "minute": 0, "day": 1}

    # =========================================================================
    # Combined Operations
    # =========================================================================

    async def apply_stat_calc_result(
        self,
        stat_changes: List[Dict[str, Any]],
        inventory_changes: List[Dict[str, Any]],
    ) -> Dict[str, Any]:
        """
        Apply results from Stat Calculator sub-agent (FS-first, then DB cache).

        This is the main entry point for mechanics_tools.py.

        Args:
            stat_changes: List of {"stat_name": str, "delta": int, "new_value": int}
            inventory_changes: List of {"action": "add"|"remove", "item_id": str, "name": str, ...}

        Returns:
            Dict with "stats_result" and "inventory_results" keys
        """
        results: Dict[str, Any] = {
            "stats_result": None,
            "inventory_results": [],
        }

        # Apply stat changes
        if stat_changes:
            changes = {sc["stat_name"]: sc["delta"] for sc in stat_changes}
            results["stats_result"] = await self.update_stats(changes)

        # Apply inventory changes
        for inv_change in inventory_changes:
            action = inv_change.get("action", "add")
            item_id = inv_change.get("item_id", "")
            name = inv_change.get("name", "")

            if not item_id:
                logger.warning(f"Skipping inventory change: missing item_id in {inv_change}")
                continue

            if action == "add":
                if not name:
                    logger.warning(f"Skipping add: missing name for item_id={item_id}")
                    continue
                result = await self.add_item(
                    item_id=item_id,
                    name=name,
                    quantity=inv_change.get("quantity", 1),
                    description=inv_change.get("description"),
                    properties=inv_change.get("properties"),
                )
            elif action == "remove":
                result = await self.remove_item(
                    item_id=item_id,
                    quantity=inv_change.get("quantity", 1),
                )
            else:
                logger.warning(f"Unknown inventory action: {action}")
                continue

            if result:
                results["inventory_results"].append(result)

        return results

File: backend-ts/src/services/player-facade.ts
/**
 * Filesystem-first player mutations, with a write-through to `player_states`.
 *
 * Ported from `backend/services/facades/player_facade.py`.
 *
 * `worlds/{name}/player.yaml` is the authoritative save file; the
 * `player_states` row is a cache the polling endpoint reads. Both have to move
 * together, and *that* is the only reason this class exists rather than a
 * handful of {@link PlayerService} calls: a stat change that lands only on disk
 * is invisible to the player until something else forces a full state reload,
 * and one that lands only in the row is lost on the next `savePlayerState`.
 *
 * ## The two sides are not symmetric
 *
 * The filesystem write is the operation. The database write is a mirror of
 * whatever the filesystem now holds, and it is allowed to fail: Python wraps
 * `_sync_to_db` in a bare `except` that logs a warning
 * (`player_facade.py:127-128`), so a locked or closed database degrades the
 * player's *display* for one turn instead of failing their turn. Reproduced
 * exactly — every method here returns success once `player.yaml` is written.
 *
 * This also means the mirror is a whole-value overwrite, not a replay of the
 * delta. The `crud/player-state.ts` mutators are deliberately *not* used for
 * it: `updateStats` there would add the same delta a second time, to a blob
 * that may already have drifted from disk, and `addInventoryItem` would store
 * the embedded item shape where this path stores what the save file holds. The
 * facade's contract is "the row says what the file says", so it writes the two
 * columns directly.
 *
 * ## Parity landmines
 *
 * - **`_sync_to_db` never inserts.** A world whose `player_states` row does not
 *   exist yet is written to disk and skipped here (`if db_state:`). Character
 *   creation is what creates that row; the facade must not race it.
 * - **The clock has no column.** `player_states` stores stats, inventory,
 *   effects and action history — there is nowhere to put `game_time`. Python
 *   still calls `_sync_to_db` after `advance_time`, so advancing the clock
 *   rewrites the *stats and inventory* columns from disk as a side effect. That
 *   is load-bearing in one direction (it repairs a row that had drifted) and
 *   surprising in the other (it overwrites whatever a `crud` mutator wrote
 *   since the last save), and it is reproduced rather than optimised away.
 * - **The inventory column ends up in mixed shapes.** What is synced is the
 *   *in-memory* inventory: entries read from `player.yaml` are in reference
 *   form (`{item_id, quantity, instance_properties}`), while an entry
 *   {@link addItem} just merged in is in the embedded form
 *   `InventoryItem.toDict()` produces. `savePlayerState` converts on the way to
 *   disk but does not touch the object it was handed, so the row receives the
 *   pre-conversion list. Python behaves identically (`player_service.py:105`
 *   builds a separate `inventory_refs`), and every reader of that column goes
 *   through `InventoryItem.fromReference`, which accepts both shapes.
 * - **The loaded state is the cached state.** {@link PlayerService} hands back
 *   the object it cached, so assigning to `state.stats` here mutates the cache
 *   too. If the filesystem write then throws, the in-memory cache is left
 *   holding values that were never persisted, until `player.yaml`'s mtime
 *   moves. Python's `_player_state_cache` has exactly this property.
 *
 * ## Binding
 *
 * Python constructs the facade per world (`PlayerFacade(world_name, db,
 * world_id)`); {@link PlayerMutationsPort} instead takes the world *name* per
 * call, so the filesystem side is chosen by the caller and only the database
 * binding lives on the instance. Build one per turn, the way `servers.ts` binds
 * a `PersistenceManager`. Constructed without `db`/`worldId` it is the
 * filesystem-only mode Python's `PlayerFacade(world_name)` gives, which is what
 * a sub-agent running before the world has a row wants.
 */

import { eq } from 'drizzle-orm'

import {
  PlayerService,
  type PlayerState,
  type StatDefinitions as FileStatDefinitions,
} from './player-service'
import type { Db } from '../db'
import { playerStates } from '../db/schema'
import {
  applyStatChanges,
  InventoryItem,
  mergeInventoryItem,
  removeInventoryItem,
  type InventoryEntry,
  type StatDefinition,
  type StatDefinitions as RuleStatDefinitions,
} from '../domain/player-rules'
import { PlayerStateSerializer } from '../domain/player-state-serializer'
import { getLogger } from '../infrastructure/logging/logger'
// Type-only, so this does not put a runtime edge from `services/` into `sdk/`:
// the interface lives next to the handlers that consume it, and implementing it
// by name is what stops the two drifting apart.
import type { PlayerMutationsPort, TimeAdvanceResult } from '../sdk/handlers/ports'

const logger = getLogger('PlayerFacade')

/** Minutes in a day — the modulus `advance_time` rolls the clock over on. */
const MINUTES_PER_DAY = 1440

/**
 * Hand `stats.yaml` to the domain rules.
 *
 * The same cast `player-service.ts` performs, and for the same reason: this
 * layer read whatever the file held, and `buildStatMap` throws on an entry with
 * no `name` exactly as Python's `stat["name"]` raises. Pre-empting that here
 * would silently leave the stat unclamped.
 */
function asRuleDefinitions(definitions: FileStatDefinitions): RuleStatDefinitions {
  return { ...definitions, stats: definitions.stats as StatDefinition[] }
}

/** Two-digit clock field, matching Python's `{:02d}` in the log line. */
function pad2(value: number): string {
  return String(value).padStart(2, '0')
}

export class PlayerFacade implements PlayerMutationsPort {
  /**
   * @param players The filesystem layer. Shared with the rest of the turn's
   *   services so they see one another's writes through one mtime cache.
   * @param db Omit for filesystem-only mode.
   * @param worldId The row to mirror onto. Required for any database write;
   *   `0` disables the sync as surely as `undefined` does, matching Python's
   *   truthiness check on `self.world_id`.
   */
  constructor(
    private readonly players: PlayerService,
    private readonly db?: Db | null,
    private readonly worldId?: number | null,
  ) {}

  // ==========================================================================
  // Write-through
  // ==========================================================================

  /**
   * Mirror the filesystem state onto the `player_states` row.
   *
   * Never throws. The caller has already written `player.yaml`, so the
   * mutation succeeded whatever happens here; all a failure costs is a stale
   * polling read until the next successful sync.
   *
   * The `SELECT` before the `UPDATE` is Python's `if db_state:` and is not
   * redundant with the `WHERE`: a bare `UPDATE` matching no rows would log a
   * sync that did not happen, and "this world has no player state yet" is a
   * routine state during onboarding rather than an anomaly.
   */
  private syncToDb(worldName: string, state: PlayerState): void {
    if (!this.db || !this.worldId) return

    try {
      const row = this.db
        .select({ id: playerStates.id })
        .from(playerStates)
        .where(eq(playerStates.worldId, this.worldId))
        .get()

      if (!row) return

      this.db
        .update(playerStates)
        .set({
          stats: PlayerStateSerializer.serializeStats(state.stats),
          inventory: PlayerStateSerializer.serializeInventory(state.inventory),
        })
        .where(eq(playerStates.worldId, this.worldId))
        .run()

      logger.debug(`📥 Synced player state to DB cache for world ${this.worldId} (${worldName})`)
    } catch (error) {
      logger.warning(`Failed to sync player state to DB: ${String(error)}`)
    }
  }

  // ==========================================================================
  // Stats
  // ==========================================================================

  /**
   * Apply `name -> delta` changes, clamped by the world's stat definitions.
   *
   * `null` means the world has no `player.yaml` — a world that does not exist,
   * or one whose files were deleted underneath a running turn. Python returns
   * `None` and logs the same warning; the tool handler treats it as "no stats
   * changed" rather than a failed turn.
   *
   * {@link PlayerService.updateStats} does load-apply-save already, but it
   * reports a missing save file as `{}`, which a world whose stats are legitimately
   * empty is indistinguishable from — and the write-through needs the loaded
   * state's *inventory* anyway. So the state is loaded here and the domain rule
   * is applied directly, which is also what Python does.
   */
  updateStats(worldName: string, changes: Record<string, number>): Record<string, number> | null {
    const state = this.players.loadPlayerState(worldName)
    if (!state) {
      logger.warning(`Player state not found for world: ${worldName}`)
      return null
    }

    state.stats = applyStatChanges(
      state.stats,
      changes,
      asRuleDefinitions(this.players.loadStatDefinitions(worldName)),
    )

    this.players.savePlayerState(worldName, state)
    logger.info(`📊 Updated stats for ${worldName}: ${JSON.stringify(changes)}`)

    this.syncToDb(worldName, state)
    return state.stats
  }

  // ==========================================================================
  // Inventory
  // ==========================================================================

  /**
   * Add an item, stacking onto an entry with the same id.
   *
   * `false` only when there is no `player.yaml` (Python's `None`). The add
   * itself cannot fail: nothing here checks that a template exists — that gate
   * lives in `change_stat`, which refuses to add an item the world has not
   * defined, while `persist_item` calls this immediately *after* writing the
   * template.
   *
   * The template is created regardless, by `savePlayerState` →
   * `ItemService.toReferenceFormat`, since the reference this writes into
   * `player.yaml` would otherwise dangle.
   */
  addItem(
    worldName: string,
    item: {
      itemId: string
      name: string
      quantity?: number
      description?: string | null
      properties?: Record<string, unknown> | null
    },
  ): boolean {
    const state = this.players.loadPlayerState(worldName)
    if (!state) {
      logger.warning(`Player state not found for world: ${worldName}`)
      return false
    }

    const quantity = item.quantity ?? 1
    state.inventory = mergeInventoryItem(
      state.inventory,
      new InventoryItem({
        id: item.itemId,
        name: item.name,
        quantity,
        description: item.description,
        properties: item.properties,
      }),
    )

    this.players.savePlayerState(worldName, state)
    logger.info(`📦 Added item to ${worldName}: ${item.name} x${quantity}`)

    this.syncToDb(worldName, state)
    return true
  }

  /**
   * Remove `quantity` of an item.
   *
   * `false` covers three cases Python's caller cannot distinguish either: no
   * `player.yaml`, no such item, and holding fewer than were asked for. The
   * third writes nothing at all — {@link removeInventoryItem} refuses a partial
   * removal, so a caller charging an item cost cannot end up having taken part
   * payment. Neither side is touched on a failure, which keeps the row and the
   * file in step through the failure too.
   *
   * Python returns the shortfall as a message on its result object; the port
   * returns a boolean, so the only place left for "you have 2, not 4" is the
   * log — the alternative is losing it entirely.
   */
  removeItem(worldName: string, itemId: string, quantity = 1): boolean {
    const state = this.players.loadPlayerState(worldName)
    if (!state) {
      logger.warning(`Player state not found for world: ${worldName}`)
      return false
    }

    const result = removeInventoryItem(state.inventory, itemId, quantity)
    if (!result.success) {
      logger.info(
        `Cannot remove ${quantity} of item ${itemId} (only ${result.remaining} available)`,
      )
      return false
    }

    state.inventory = result.inventory

    this.players.savePlayerState(worldName, state)
    logger.info(`📦 Removed item from ${worldName}: ${itemId} x${quantity}`)

    this.syncToDb(worldName, state)
    return true
  }

  /**
   * The inventory with every reference resolved against its `items/` template.
   *
   * Python's `get_inventory(resolved=True)`. The `resolved=False` branch is not
   * ported: {@link PlayerMutationsPort} does not offer it, and the only Python
   * caller (`mechanics_tools.py:233`) asks for the resolved list — a raw
   * reference list would render as bare ids in the model's context.
   */
  getInventory(worldName: string): InventoryEntry[] {
    return this.players.getResolvedInventory(worldName)
  }

  // ==========================================================================
  // Clock
  // ==========================================================================

  /**
   * Advance the in-world clock, rolling over at 24:00.
   *
   * `null` for a non-positive `minutes` — checked *before* the state is loaded,
   * so "advance by zero" is not even a read — and for a world with no
   * `player.yaml`. `advance_time` renders its success text off this result, so
   * both cases silently degrade to a message without a time in it.
   *
   * There is no clock column to sync; see the note on the write-through in this
   * file's header for why the sync is issued anyway.
   */
  advanceTime(worldName: string, minutes: number): TimeAdvanceResult | null {
    if (minutes <= 0) return null

    const state = this.players.loadPlayerState(worldName)
    if (!state) {
      logger.warning(`Player state not found for world: ${worldName}`)
      return null
    }

    const oldTime = { ...state.gameTime }

    // Python's `//` and `%` floor toward negative infinity and take the
    // divisor's sign; JS's `/` and `%` do neither. Subtracting the floored
    // quotient reproduces Python's pair for a clock that has somehow gone
    // negative, which is the only input where the two disagree.
    const totalMinutes = state.gameTime.hour * 60 + state.gameTime.minute + minutes
    const daysPassed = Math.floor(totalMinutes / MINUTES_PER_DAY)
    const remainingMinutes = totalMinutes - daysPassed * MINUTES_PER_DAY

    const newHour = Math.floor(remainingMinutes / 60)
    state.gameTime = {
      hour: newHour,
      minute: remainingMinutes - newHour * 60,
      day: state.gameTime.day + daysPassed,
    }

    this.players.savePlayerState(worldName, state)
    logger.info(
      `⏰ Advanced time for ${worldName}: +${minutes}min -> ` +
        `${pad2(state.gameTime.hour)}:${pad2(state.gameTime.minute)} Day ${state.gameTime.day}`,
    )

    this.syncToDb(worldName, state)
    return { oldTime, newTime: state.gameTime }
  }

  // ==========================================================================
  // Reads
  // ==========================================================================

  /**
   * The raw save state, for the handlers that need the clock or the turn count.
   *
   * A straight delegation, present because the handlers hold the port rather
   * than a {@link PlayerService}. Python's equivalents are the facade's
   * `get_stats` / `get_game_time`; those two, and `apply_stat_calc_result`,
   * have no call site in the Python tree and are not ported.
   */
  loadPlayerState(worldName: string): PlayerState | null {
    return this.players.loadPlayerState(worldName)
  }
}
